File: gc-commons/gc-base-common/src/main/java/com/gc/common/base/spring/ApplicationContextRegister.java
package com.gc.common.base.spring;

import lombok.extern.log4j.Log4j2;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Component;

/**
 * 容器注册类
 * @author jackson
 */
@Component
@Log4j2
public class ApplicationContextRegister implements ApplicationContextAware {

    private static ApplicationContext applicationContext;

    /**
     * 通过名字获取bean
     * @param name
     * @return
     */
    @Nullable
    public static Object getBean(@NonNull String name) {
        try {
            return applicationContext.getBean(name);
        } catch (Exception e) {
            log.error("获取bean失败", e);
            return null;
        }
    }

    /**
     * 通过类型获取bean
     * @param clazz
     * @param <T>
     * @return
     */
    @Nullable
    public static <T> T getBean(@NonNull Class<T> clazz) {
        try {
            return applicationContext.getBean(clazz);
        } catch (Exception e) {
            log.error("获取bean失败", e);
            return null;
        }
    }

    /**
     * 获取spring上下文
     * @return
     */
    @NonNull
    public static ApplicationContext getContext() {
        return applicationContext;
    }


    @Override
    public void setApplicationContext(@NonNull ApplicationContext applicationContext) {
        log.debug("ApplicationContext registed-->{}", applicationContext);
        ApplicationContextRegister.applicationContext = applicationContext;
    }
}

File: smart-commons/smart-common/src/main/kotlin/com/smart/common/spring/ApplicationContextRegister.kt
package com.smart.common.spring

import org.slf4j.LoggerFactory
import org.springframework.beans.BeansException
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.ApplicationContext
import org.springframework.context.ApplicationContextAware
import org.springframework.stereotype.Component

/**
 * 容器注册类
 * @author ming
 * 2019/7/6 下午8:24
 */
@ConditionalOnMissingBean(ApplicationContextRegister::class)
@Component
class ApplicationContextRegister : ApplicationContextAware {


    companion object {
        private val LOGGER = LoggerFactory.getLogger(ApplicationContextRegister::class.java)
        private lateinit var applicationContext: ApplicationContext

        /**
         * 通过名字获取bean
         * @param name
         * @return
         */
        fun getBean(name: String): Any? {
            return applicationContext.getBean(name)
        }

        fun <T> getBean(type: Class<T>): T? {
            return applicationContext.getBean(type)
        }

        fun getContext(): ApplicationContext {
            return applicationContext
        }
    }

    @Throws(BeansException::class)
    override fun setApplicationContext(applicationContext: ApplicationContext) {
        LOGGER.debug("ApplicationContext registed-->{}", applicationContext)
        Companion.applicationContext = applicationContext
    }


}